File: matlab/run_tournament.m
%RUN_TOURNAMENT  Presentation tournament runner: heuristic vs mcts, round-robin.
%
%   Saves timestamped outputs to matlab/outputs/:
%     YYYYMMDD_HHMMSS_win_rates.csv
%     YYYYMMDD_HHMMSS_win_rate_chart.png
%     YYYYMMDD_HHMMSS_avg_game_length.csv
%     YYYYMMDD_HHMMSS_vp_final.csv
%
%   Edit PARAMS below, then run:  run_tournament

%% ── PARAMS ───────────────────────────────────────────────────────────────────
PARAMS.agents  = {'heuristic', 'mcts'};
PARAMS.N       = 50;   % games per ordered matchup
PARAMS.rngSeed = 1;    % base seed (incremented per game for variety)

%% ── Setup ────────────────────────────────────────────────────────────────────
ts        = datestr(now, 'yyyymmdd_HHMMSS');  %#ok<TNOW1,DATST>
outputDir = fullfile(fileparts(mfilename('fullpath')), 'outputs');
if ~exist(outputDir, 'dir'), mkdir(outputDir); end

config                = catan_core('defaultConfig');
config.showViz        = false;
config.pauseAfterMove = false;
config.verbose        = false;

agentNames = PARAMS.agents;
N          = PARAMS.N;
nAgents    = numel(agentNames);

agentFnMap = containers.Map( ...
    {'random','heuristic','monte_carlo','mcts'}, ...
    {@agent_random, @agent_heuristic, @agent_montecarlo, @agent_mcts});

% Accumulators
wins        = zeros(1, nAgents);
losses      = zeros(1, nAgents);
gameLengths = [];
vpByAgent   = cell(1, nAgents);   % final VP per game per agent

% All ordered pairs (i,j) where i ~= j
matchups = [];
for ii = 1:nAgents
    for jj = 1:nAgents
        if ii ~= jj
            matchups(end+1,:) = [ii, jj]; %#ok<AGROW>
        end
    end
end

fprintf('\n%s\n', repmat('=',1,54));
fprintf('  TOURNAMENT  (%d games per ordered matchup)\n', N);
fprintf('  Agents: %s\n', strjoin(agentNames, ', '));
fprintf('%s\n', repmat('=',1,54));

baseRngSeed = PARAMS.rngSeed;
gameNum     = 0;

%% ── Round-robin ──────────────────────────────────────────────────────────────
for m = 1:size(matchups,1)
    iA     = matchups(m,1);
    iB     = matchups(m,2);
    p1Name = agentNames{iA};
    p2Name = agentNames{iB};
    p1Fn   = agentFnMap(p1Name);
    p2Fn   = agentFnMap(p2Name);

    fprintf('  %-15s (P1)  vs  %-15s (P2) ...', p1Name, p2Name);
    p1Wins = 0;

    for g = 1:N
        gameNum        = gameNum + 1;
        cfg            = config;
        cfg.rngSeed    = baseRngSeed + gameNum;

        history = catan_core('simulateGame', {p1Fn, p2Fn}, cfg);
        fs      = history.finalState;

        if fs.winnerId == 1
            wins(iA)   = wins(iA)   + 1;
            losses(iB) = losses(iB) + 1;
            p1Wins     = p1Wins + 1;
        elseif fs.winnerId == 2
            wins(iB)   = wins(iB)   + 1;
            losses(iA) = losses(iA) + 1;
        end

        gameLengths(end+1)     = fs.turnIndex; %#ok<AGROW>
        vpByAgent{iA}(end+1)   = fs.players(1).victoryPoints;
        vpByAgent{iB}(end+1)   = fs.players(2).victoryPoints;
    end

    fprintf('  %d/%d for P1\n', p1Wins, N);
end

%% ── Console summary ──────────────────────────────────────────────────────────
avgLen = mean(gameLengths);

fprintf('\n%s\n', repmat('=',1,54));
fprintf('  RESULTS\n');
fprintf('%s\n', repmat('=',1,54));
for i = 1:nAgents
    total = wins(i) + losses(i);
    rate  = wins(i) / max(total, 1);
    avgVP = mean(vpByAgent{i});
    fprintf('  %-15s: %dW / %dL  (%.1f%% win rate)  avg VP: %.2f\n', ...
        agentNames{i}, wins(i), losses(i), rate*100, avgVP);
end
fprintf('  Avg game length : %.1f turns\n', avgLen);
fprintf('  Total games     : %d\n\n', numel(gameLengths));

%% ── win_rates.csv ────────────────────────────────────────────────────────────
fid = fopen(fullfile(outputDir, sprintf('%s_win_rates.csv', ts)), 'w');
fprintf(fid, 'agent,wins,losses,win_rate\n');
for i = 1:nAgents
    total = wins(i) + losses(i);
    rate  = wins(i) / max(total,1);
    fprintf(fid, '%s,%d,%d,%.4f\n', agentNames{i}, wins(i), losses(i), rate);
end
fclose(fid);

%% ── win_rate_chart.png ───────────────────────────────────────────────────────
rates = zeros(1, nAgents);
for i = 1:nAgents
    total    = wins(i) + losses(i);
    rates(i) = wins(i) / max(total, 1);
end

% Muted palette: blue, orange, green, gold
palette = [0.357 0.608 0.835;
           0.929 0.490 0.192;
           0.439 0.678 0.278;
           1.000 0.753 0.000];

fig1 = figure('Visible','off','Position',[0 0 520 440]);
ax1  = axes(fig1);
b    = bar(ax1, 1:nAgents, rates, 0.5);
b.FaceColor = 'flat';
for i = 1:nAgents
    b.CData(i,:) = palette(mod(i-1, size(palette,1))+1, :);
end
ax1.XTick      = 1:nAgents;
ax1.XTickLabel = agentNames;
ax1.YLim       = [0, 1.15];
ax1.YGrid      = 'on';
ax1.XGrid      = 'off';
ax1.GridLineStyle = '--';
ax1.GridAlpha  = 0.4;
ax1.TickLength = [0, 0];
box(ax1, 'off');
ylabel(ax1, 'Win Rate');
title(ax1, sprintf('Agent Win Rates  (N=%d total games)', numel(gameLengths)));
for i = 1:nAgents
    text(ax1, i, rates(i) + 0.03, sprintf('%.1f%%', rates(i)*100), ...
        'HorizontalAlignment','center','FontSize',11,'FontWeight','bold');
end
print(fig1, fullfile(outputDir, sprintf('%s_win_rate_chart.png', ts)), '-dpng', '-r150');
close(fig1);

%% ── avg_game_length.csv ──────────────────────────────────────────────────────
fid = fopen(fullfile(outputDir, sprintf('%s_avg_game_length.csv', ts)), 'w');
fprintf(fid, 'metric,value\n');
fprintf(fid, 'avg_turns,%.2f\n', avgLen);
fprintf(fid, 'min_turns,%d\n',   min(gameLengths));
fprintf(fid, 'max_turns,%d\n',   max(gameLengths));
fprintf(fid, 'total_games,%d\n', numel(gameLengths));
fclose(fid);

%% ── vp_final.csv ─────────────────────────────────────────────────────────────
fid = fopen(fullfile(outputDir, sprintf('%s_vp_final.csv', ts)), 'w');
fprintf(fid, 'agent,avg_vp\n');
for i = 1:nAgents
    fprintf(fid, '%s,%.4f\n', agentNames{i}, mean(vpByAgent{i}));
end
fclose(fid);

fprintf('  Outputs saved to: %s\n', outputDir);
fprintf('  Timestamp prefix: %s\n\n', ts);

File: run_tournament.py
"""
run_tournament.py — Presentation tournament runner.

Round-robin between agents (default: heuristic vs mcts).
Saves timestamped outputs to outputs/:
  YYYYMMDD_HHMMSS_win_rates.csv
  YYYYMMDD_HHMMSS_win_rate_chart.png
  YYYYMMDD_HHMMSS_avg_game_length.csv
  YYYYMMDD_HHMMSS_vp_final.csv

Usage
-----
  python run_tournament.py                            # heuristic vs mcts, 50 games
  python run_tournament.py --agents heuristic mcts --n 20
"""

import argparse
import csv
import os
from datetime import datetime

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from config import default_config
from agents import resolve_agent
from game import simulate_game

# ── Defaults ──────────────────────────────────────────────────────────────────

DEFAULT_AGENTS = ["heuristic", "mcts"]
DEFAULT_N      = 50
OUTPUT_DIR     = "outputs"
COLORS         = ["#5B9BD5", "#ED7D31", "#70AD47", "#FFC000"]


# ── Core runner ───────────────────────────────────────────────────────────────

def run_presentation_tournament(
    agent_names: list = None,
    N: int = DEFAULT_N,
    config: dict = None,
) -> dict:
    """
    Run a round-robin tournament and save presentation outputs.

    Parameters
    ----------
    agent_names : list of str
        Agent type strings (e.g. ['heuristic', 'mcts']).
    N : int
        Games per ordered matchup.
    config : dict, optional
        Base config; verbose/rng_seed used, viz/pause ignored.

    Returns
    -------
    dict with wins, losses, game_lengths, vp_by_agent
    """
    if agent_names is None:
        agent_names = DEFAULT_AGENTS
    if config is None:
        config = default_config()

    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    cfg                = dict(config)
    cfg["verbose"]     = False
    cfg["num_players"] = 2

    # Ordered pairs where a != b
    matchups = [(a, b) for a in agent_names for b in agent_names if a != b]

    # Accumulators
    wins        = {n: 0 for n in agent_names}
    losses      = {n: 0 for n in agent_names}
    game_lengths: list = []
    vp_by_agent = {n: [] for n in agent_names}

    base_seed = cfg.get("rng_seed", 42)
    game_num  = 0

    print(f"\n{'=' * 54}")
    print(f"  TOURNAMENT  ({N} games per ordered matchup)")
    print(f"  Agents: {', '.join(agent_names)}")
    print(f"{'=' * 54}")

    for p1_name, p2_name in matchups:
        p1_wins = 0
        print(f"  {p1_name:<15} (P1)  vs  {p2_name:<15} (P2) ...",
              end="", flush=True)

        for _ in range(N):
            game_num            += 1
            game_cfg             = dict(cfg)
            game_cfg["rng_seed"] = base_seed + game_num

            history = simulate_game(
                [resolve_agent(p1_name), resolve_agent(p2_name)],
                game_cfg,
                player_names=[p1_name, p2_name],
            )
            state = history["final_state"]

            if state.winner_id == 1:
                wins[p1_name]   += 1
                losses[p2_name] += 1
                p1_wins         += 1
            elif state.winner_id == 2:
                wins[p2_name]   += 1
                losses[p1_name] += 1

            game_lengths.append(state.turn_index)
            vp_by_agent[p1_name].append(state.players[1].victory_points)
            vp_by_agent[p2_name].append(state.players[2].victory_points)

        print(f"  {p1_wins}/{N} for P1")

    # ── Console summary ───────────────────────────────────────────────────────

    avg_len = sum(game_lengths) / len(game_lengths)

    print(f"\n{'=' * 54}")
    print("  RESULTS")
    print(f"{'=' * 54}")
    for name in agent_names:
        total   = wins[name] + losses[name]
        rate    = wins[name] / total if total > 0 else 0.0
        avg_vp  = sum(vp_by_agent[name]) / len(vp_by_agent[name])
        print(f"  {name:<15}: {wins[name]}W / {losses[name]}L  "
              f"({rate:.1%} win rate)  avg VP: {avg_vp:.2f}")
    print(f"  Avg game length : {avg_len:.1f} turns")
    print(f"  Total games     : {len(game_lengths)}")

    # ── win_rates.csv ─────────────────────────────────────────────────────────

    wr_path = os.path.join(OUTPUT_DIR, f"{ts}_win_rates.csv")
    with open(wr_path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["agent", "wins", "losses", "win_rate"])
        for name in agent_names:
            total = wins[name] + losses[name]
            rate  = wins[name] / total if total > 0 else 0.0
            w.writerow([name, wins[name], losses[name], f"{rate:.4f}"])

    # ── win_rate_chart.png ────────────────────────────────────────────────────

    rates  = []
    for name in agent_names:
        total = wins[name] + losses[name]
        rates.append(wins[name] / total if total > 0 else 0.0)

    colors = COLORS[:len(agent_names)]
    fig, ax = plt.subplots(figsize=(6, 5))
    bars = ax.bar(agent_names, rates, color=colors, width=0.5)

    for bar, rate in zip(bars, rates):
        ax.text(
            bar.get_x() + bar.get_width() / 2,
            bar.get_height() + 0.025,
            f"{rate:.1%}",
            ha="center", va="bottom", fontsize=11, fontweight="bold",
        )

    total_games = len(matchups) * N
    ax.set_ylabel("Win Rate")
    ax.set_title(f"Agent Win Rates  (N={total_games} total games)")
    ax.set_ylim(0, 1.15)
    ax.yaxis.grid(True, linestyle="--", alpha=0.4)
    ax.xaxis.grid(False)
    ax.set_axisbelow(True)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.tick_params(axis="x", bottom=False)
    plt.tight_layout()

    chart_path = os.path.join(OUTPUT_DIR, f"{ts}_win_rate_chart.png")
    plt.savefig(chart_path, dpi=150, bbox_inches="tight")
    plt.close()

    # ── avg_game_length.csv ───────────────────────────────────────────────────

    gl_path = os.path.join(OUTPUT_DIR, f"{ts}_avg_game_length.csv")
    with open(gl_path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["metric", "value"])
        w.writerow(["avg_turns",   f"{avg_len:.2f}"])
        w.writerow(["min_turns",   min(game_lengths)])
        w.writerow(["max_turns",   max(game_lengths)])
        w.writerow(["total_games", len(game_lengths)])

    # ── vp_final.csv ──────────────────────────────────────────────────────────

    vp_path = os.path.join(OUTPUT_DIR, f"{ts}_vp_final.csv")
    with open(vp_path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["agent", "avg_vp"])
        for name in agent_names:
            avg_vp = sum(vp_by_agent[name]) / len(vp_by_agent[name])
            w.writerow([name, f"{avg_vp:.4f}"])

    print(f"\n  Outputs saved to {OUTPUT_DIR}/{ts}_*.{{csv,png}}")

    return {
        "wins":         wins,
        "losses":       losses,
        "game_lengths": game_lengths,
        "vp_by_agent":  vp_by_agent,
    }


# ── CLI ───────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Presentation tournament runner.")
    parser.add_argument(
        "--agents", nargs="+", default=DEFAULT_AGENTS,
        help="Agent names (default: heuristic mcts)",
    )
    parser.add_argument(
        "--n", type=int, default=DEFAULT_N,
        help="Games per ordered matchup (default: 50)",
    )
    args = parser.parse_args()
    run_presentation_tournament(agent_names=args.agents, N=args.n)
